engineapi/proxy.go

```diff
@@ -19,77 +23,330 @@ import (
 )
 
 const (
-	ProxyPollInterval = 5 * time.Second
+	MaxProxyClientConcurrentInUse = 1000
+	DefaultProxyClientPerImage    = 3
+	MaxProxyClientPerImage        = 3 * DefaultProxyClientPerImage
+	ProxyPollInterval             = 5 * time.Second
+	ProxyClientMaxRetry           = 10
+	ProxyGCMaxRetry               = 3
 )
 
-type EngineClientProxy interface {
-	EngineClient
-	BackupTargetBinaryClient
+type EngineClientProxyManager struct {
+	handler *EngineClientProxyHandler
 
-	IsGRPC() bool
-	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
-	Stop(string) error
-	Ping() error
+	ds     *datastore.DataStore
+	logger logrus.FieldLogger
+
+	stopCh       chan struct{}
+	lock         *sync.RWMutex
+	clock        clock.Clock
+	resyncPeriod time.Duration
+}
+
+func NewEngineClientProxyManager(logger logrus.FieldLogger, ds *datastore.DataStore, resyncPeriod time.Duration, stopCh chan struct{}) *EngineClientProxyManager {
+	return &EngineClientProxyManager{
+		handler:      NewEngineClientProxyHandler(logger, ds, stopCh),
+		logger:       logger,
+		ds:           ds,
+		resyncPeriod: resyncPeriod,
+		clock:        &clock.RealClock{},
+		lock:         &sync.RWMutex{},
+		stopCh:       stopCh,
+	}
+}
+
+func (m *EngineClientProxyManager) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	m.logger.Info("Start Longhorn proxy manager")
+	defer m.logger.Info("Shutting down Longhorn proxy manager")
+
+	go wait.Until(m.resync, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func getLoggerForProxyManager(logger logrus.FieldLogger, imName string) *logrus.Entry {
+	return logger.WithField("instanceManager", imName)
+}
+
+func (m *EngineClientProxyManager) resync() {
```

Comment:
One quick question, is this layer/abstraction necessary?
It seems that this resync can be done by the ProxyHandler as well.

Reply:
Fixed.